### ISSUE TITLE
Reproducibility of results without AVX512 by setting ATEN_CPU_CAPABILITY=avx2

### ISSUE THREAD

**Issue by TimotheeMathieu** [pytorch/pytorch#155552]
### 🐛 Describe the bug

I get different results when executing the following minimal example on a computer with AVX-512 compared to a computer without AVX-512, even when `ATEN_CPU_CAPABILITY=avx2`

```
import numpy as np
import torch
import torch.nn as nn
import random
torch.use_deterministic_algorithms(True)

seed = 42

random.seed(seed)
torch.manual_seed(seed)
np.random.seed(seed)

device = torch.device("cpu")

def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
    torch.nn.init.orthogonal_(layer.weight, std)
    torch.nn.init.constant_(layer.bias, bias_const)
    return layer

network = nn.Sequential(layer_init(nn.Linear(100, 100)),
                        layer_init(nn.Linear(100,1), std=1.0),).to(device)
with torch.no_grad():
    action = network(torch.rand(size=(100,)).to(device))

print(format(action.cpu().numpy()[0].item(), '.60g'))
```

To be sure that it was not a hardware or environment problem I tried the above in a vm constructed through qemu in which I removed the avx512f flag. The experiment went as follow

- Construct a qemu vm with pytorch setup
- launch the qemu vm with argument `-cpu host` and execute `ATEN_CPU_CAPABILITY=avx2 python3 test_torch.py`, record the result
- launch the qemu vm with argument `-cpu host,-avx512f` and execute `ATEN_CPU_CAPABILITY=avx2 python3 test_torch.py`, record the result

The results are not the same. Result on AVX512 vm is 0.4804637432098388671875 on non-AVX512 vm I get 0.4804628789424896240234375, a difference of order 1e-5. Remark that this discrepency is also visible when doing the test on two different computers but executing in vm was convenient to test without any doubt that the problem could be a hardware problem.

### Versions

PyTorch version: 2.5.0a0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: linux (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.7 (main, Jan  1 1970, 00:00:01) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.12.31-1-lts-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
/gnu/store/m0xdsa8cfq6mq1kxgxmpmpg71la4f0b9-bash-minimal-5.1.16/bin/sh: line 1: lscpu: command not found

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] onnx==1.17.0
[pip3] onnxoptimizer==0.3.19
[pip3] optree==0.14.0
[pip3] torch==2.5.0a0+gitunknown
[conda] Could not collect

cc @malfet

**Comment by malfet** [comment]
nn.Linear dispatches directly to MKL, which does not respect ATEN settings... But failures are within the accepted tolerances

**Comment by TimotheeMathieu** [comment]
Thanks, I tried to  disable AVX512 in both ATEN and MKL, but sadly I still get different results on the vm with AVX512 flag and the one without. This means that AVX512 is still used somewhere that is neither in ATEN or in MKL although I don't know where.
The command I used is
```
ATEN_CPU_CAPABILITY=avx2 MKL_ENABLE_INSTRUCTIONS=AVX2 python3 test_torch.py
```
Did I use this right? I think this should have disabled AVX512 in MKL. Does someone have an idea of where AVX512 is used and how I could disable it without having to do a vm (which is slow)?

**Comment by ZhaoqiongZ** [comment]
Hi @TimotheeMathieu , could you verify on a physical machine rather than a VM, since CPUID in a VM is emulated and uncontrollable.

**Comment by TimotheeMathieu** [comment]
@ZhaoqiongZ: As explained in my first post, I did try on physical machine, in fact I first saw the discrepancy on physical machines. 

And about the VM, I use qemu with `-cpu host` which does not emulate the CPUID as it uses the host cpu (from what I understood, I am not super familiar with qemu and may be wrong about this).

**Comment by xuhancn** [comment]
Hi @TimotheeMathieu 

I tried in my physical machine, and I didn't find any discrepancy:

<img width="770" alt="Image" src="https://github.com/user-attachments/assets/267cc450-6c2b-4a93-83eb-f2585d5ae8bb" />

My CPU is fully support `AVX2` and `AVX512`.

Additional info:
1. I didn't see your CPU info, what CPU SKU of your machine?
2. We didn't guarantee the behavior in VM, because of the all CPU behavior maybe simulated by hypervisor.

**Comment by TimotheeMathieu** [comment]
@xuhancn yes I know, I have the same, there should be a discrepancy because AVX512 has a discrepancy.

If you try on two physical machines one with AVX512 and the other without, then you see that there is a difference. So my point is that `ATEN_CPU_CAPABILITY=avx2` **is not sufficient** to disable avx512 on avx512-capable machine for pytorch. As @ZhaoqiongZ was saying it must be because AVX512 is used by something other than ATEN.

**Comment by xuhancn** [comment]
@TimotheeMathieu 

I tested in another physical machine:
<img width="703" alt="Image" src="https://github.com/user-attachments/assets/13405c3d-d1fc-48c5-ba67-5d91ecaf0444" />

It's value is changed, I guess the value is only related to machine, but not related to AVX2/AVX512.

For VM, you changed its configuration maybe make it to be another virtual machine.

**Comment by TimotheeMathieu** [comment]
I don't know how and why, I guess you are on windows which changes a lot the libraries and how they are used so I can't really compare as I used only machines on Linux.

I don't really want to use vm at all if possible, it was just to show what the problem was but as I said I also tried on different physical machines. I tried with a total of 8 different machines and the difference seemed to be AVX-512.

**Comment by xuhancn** [comment]
Hi @TimotheeMathieu 

I tested on the Linux machine, and the result as same as on Windows:

![Image](https://github.com/user-attachments/assets/aafbacd4-08ee-4968-a43f-6c0c3caeb94b)

What CPU you are using?

**Comment by TimotheeMathieu** [comment]
@xuhancn This result is what I expect on AVX512 computer, would you mind testing on an AVX2 linux computer?

I tried on quite a lot of computers, 
Non AVX512 CPU:  Intel(R) Core(TM) i9-9980HK, 13th Gen Intel(R) Core(TM) i9-13950HX, Intel(R) Xeon(R) CPU E5-2623,  Intel(R) Xeon(R) CPU E5-2640, 
AVX 512 CPU: Intel(R) Xeon(R) Gold 6134, 11th Gen Intel(R) Core(TM) i7-1185G7, Intel(R) Xeon(R) Silver 4114 CPU

Moreover, recently I managed to use https://github.com/twosigma/libvirtcpuid/ to mask the AVX512 flag and recover the non AVX512 results on all these computers (in effect this solves my initial reproducibility problem). This shows I think conclusively that the discrepancy is indeed due to AVX512 and not due to hardware.

**Comment by xuhancn** [comment]
Hi @TimotheeMathieu 

I also tested Linux AVX2 physical machine, the result as same as Windows:

<img width="785" alt="Image" src="https://github.com/user-attachments/assets/79722a9b-900b-47da-8abf-04669222ce08" />

For `libvirtcpuid`, @ZhaoqiongZ metioned https://github.com/pytorch/pytorch/issues/155552#issuecomment-2964729324 VM is simulated CPUID, and it is not safe to simulate `AVX512f` on `AVX2` virtual CPUs.

We didin't have enough knowledge to talk about VM detaileds.

**Comment by TimotheeMathieu** [comment]
I don't simulate AVX512f, I simulate not having it but maybe you are right and this is not safe. 

I don't understand why you cannot reproduce then, as said previously I did not try only on vm, I also see the difference in physical computers. Did you use the same pytorch version and python version as the ones I used? (cf first message)

**Comment by xuhancn** [comment]
@TimotheeMathieu All of my tests are using:
torch==2.5
numpy==1.24.4
